File: Algorithms/Merge Sort/mergesort.go
//implement merge sort in go with descriptive comments

package main

func mergeSort(arr []int) []int {
	if len(arr) <= 1 { //if the length of the array is 1 or 0, then it is already sorted
		return arr
	}

	mid := len(arr) / 2 //find the middle index of the array
	left := arr[:mid]    //create a slice of the left side of the array
	right := arr[mid:]   //create a slice of the right side of the array

	return merge(
		mergeSort(left),   //recursively call mergeSort on the left side
		mergeSort(right),  //recursively call mergeSort on the right side
	)
}

func merge(left, right []int) []int {
	results := []int{} //create an empty array to store our results

	for len(left) > 0 && len(right) > 0 { //while there are still elements in both slices
		if left[0] < right[0] { //if the first element in the left slice is less than the first element in the right slice
			results = append(results, left[0]) //append the first element in the left slice to the results
			left = left[1:]                    //remove the first element from the left slice
		} else { //if the first element in the right slice is less than or equal to the first element in the left slice
			results = append(results, right[0]) //append the first element in the right slice to the results
			right = right[1:]                   //remove the first element from the right slice
		}
	}

	//if there are still elements in the left slice, append them all to the results
	for i := 0; i < len(left); i++ {
		results = append(results, left[i])
	}

	//if there are still elements in the right slice, append them all to the results
	for i := 0; i < len(right); i++ {
		results = append(results, right[i])
	}

	return results //return the sorted array
}

File: Algorithms/Merge Sort/mergesort.py
def merge_sort(arr):
    # Base case - if the input array is only one element, return it
    if len(arr) == 1:
        return arr

    # Split the input array in half
    left = arr[:len(arr) // 2]
    right = arr[len(arr) // 2:]

    # Recursively call merge_sort on the left and right halves of the array
    left = merge_sort(left)
    right = merge_sort(right)

    # Initialize an empty array to store the merged array
    merged = []

    # While there are elements in both the left and right arrays:
    while len(left) > 0 and len(right) > 0:
        # If the first element in the left array is smaller than the first element in the right array:
        if left[0] < right[0]:
            # Append the first element in the left array to the merged array
            merged.append(left[0])
            # Remove the first element in the left array
            left = left[1:]
        # Otherwise:
        else:
            # Append the first element in the right array to the merged array
            merged.append(right[0])
            # Remove the first element in the right array
            right = right[1:]

    # If there are elements remaining in the left array:
    while len(left) > 0:
        # Append the first element in the left array to the merged array
        merged.append(left[0])
        # Remove the first element in the left array
        left = left[1:]

    # If there are elements remaining in the right array:
    while len(right) > 0:
        # Append the first element in the right array to the merged array
        merged.append(right[0])
        # Remove the first element in the right array
        right = right[1:]

    # Return the merged array
    return merged